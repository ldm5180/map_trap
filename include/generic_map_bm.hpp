#ifndef GENERIC_MAP_BM_HPP
#define GENERIC_MAP_BM_HPP

#include "add_to_map.hpp"

#include <boost/hana.hpp>
#include <cstdlib>

static constexpr auto has_reserve =
    boost::hana::is_valid([](auto &&x) -> decltype((void)x.reserve(0)) {});

template <class map_t> static void BM_map_insert(benchmark::State &state) {
  std::srand(42);
  map_t map{};
  if
    constexpr(has_reserve(map)) { map.reserve(state.range(0)); }

  for (auto _ : state) {
    add_to_map(state.range(0), map);
  }
}

template <class map_t> static void BM_map_iterate(benchmark::State &state) {
  std::srand(42);
  map_t map{};
  if
    constexpr(has_reserve(map)) { map.reserve(state.range(0)); }

  for (auto i = 0; i < state.range(0); ++i) {
    add_to_map(state.range(0), map);
  }

  for (auto _ : state) {
    // Nothing to do to sort them. Can simply iterate.
    typename map_t::mapped_type v{};
    for (const auto &kv : map) {
      benchmark::DoNotOptimize(v = kv.second);
    }
  }
}

template <class map_t> static void BM_map_find(benchmark::State &state) {
  std::srand(42);
  map_t map{};
  if
    constexpr(has_reserve(map)) { map.reserve(state.range(0)); }
  std::vector<typename map_t::key_type> keys{};
  keys.reserve(state.range(0));

  for (auto i = 0; i < state.range(0); ++i) {
    add_to_map_and_key_list(state.range(0), map, keys);
  }

  for (auto _ : state) {
    typename map_t::mapped_type v{};
    const auto key = keys[std::rand() % keys.size()];
    benchmark::DoNotOptimize(v = map[key]);
  }
}

#endif // GENERIC_MAP_BM_HPP
