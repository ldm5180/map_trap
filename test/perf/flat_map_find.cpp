#include "generic_map_bm.hpp"

#include <benchmark/benchmark.h>
#include <boost/container/flat_map.hpp>

BENCHMARK_TEMPLATE(BM_map_find, boost::container::flat_map<int, std::string>)
    ->RangeMultiplier(2)
    ->Range(8, 8 << 4);
BENCHMARK_MAIN();
