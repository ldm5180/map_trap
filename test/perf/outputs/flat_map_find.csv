Run on (4 X 2594.09 MHz CPU s)
2017-10-28 17:23:54
name,iterations,real_time,cpu_time,time_unit,bytes_per_second,items_per_second,label,error_occurred,error_message
"BM_map_find<boost::container::flat_map<int64_t, int64_t>>/8",16404376,43.1431,43.1409,ns,,,,,
"BM_map_find<boost::container::flat_map<int64_t, int64_t>>/16",13504267,50.5258,50.5243,ns,,,,,
"BM_map_find<boost::container::flat_map<int64_t, int64_t>>/32",12260005,58.2788,58.2758,ns,,,,,
"BM_map_find<boost::container::flat_map<int64_t, int64_t>>/64",8382278,77.4923,77.4887,ns,,,,,
"BM_map_find<boost::container::flat_map<int64_t, int64_t>>/128",6385280,109.544,109.54,ns,,,,,
