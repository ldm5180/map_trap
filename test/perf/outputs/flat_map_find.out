Run on (4 X 2594.09 MHz CPU s)
2017-10-28 16:29:21
name,iterations,real_time,cpu_time,time_unit,bytes_per_second,items_per_second,label,error_occurred,error_message
"BM_map_find<boost::container::flat_map<int, std::string>>/8",13324564,47.1847,47.2061,ns,,,,,
"BM_map_find<boost::container::flat_map<int, std::string>>/16",13487669,50.572,50.5905,ns,,,,,
"BM_map_find<boost::container::flat_map<int, std::string>>/32",10349341,59.5437,59.5415,ns,,,,,
"BM_map_find<boost::container::flat_map<int, std::string>>/64",9065180,83.2853,83.3252,ns,,,,,
"BM_map_find<boost::container::flat_map<int, std::string>>/128",5935698,118.051,118.105,ns,,,,,
